curvefs/src/tools/list/curvefs_topology_list.cpp

```diff
@@ -0,0 +1,318 @@
+                             << " can't be found." << std::endl;
+                ret = false;
+            }
+        }
+    }
+    return ret;
+}
+
+bool TopologyListTool::GetServerInfoFromResponse() {
+    bool ret = true;
+    auto const& servers = response_->servers();
+    if (servers.statuscode() != curvefs::mds::topology::TOPO_OK) {
+        std::cerr << "the servers of cluster has error, error code is "
+                  << servers.statuscode() << ", code name is "
+                  << curvefs::mds::topology::TopoStatusCode_Name(
+                         servers.statuscode())
+                  << std::endl;
+        ret = false;
+    } else {
+        for (auto const& i : servers.serverinfos()) {
+            serverId2ServerInfo.insert(
+                std::pair<mds::topology::ServerIdType, ServerInfoType>(
+                    i.serverid(),
+                    ServerInfoType(
+                        i, std::vector<mds::topology::MetaServerIdType>())));
+
+            if (zoneId2ZoneInfo.find(i.zoneid()) != zoneId2ZoneInfo.end()) {
+                zoneId2ZoneInfo[i.zoneid()].second.emplace_back(i.serverid());
+            } else {
+                errorOutput_ << "server:" << i.serverid()
+                             << " has error: zoneId:" << i.zoneid()
+                             << " can't be found." << std::endl;
+                ret = false;
+            }
+        }
+    }
+    return ret;
+}
+
+bool TopologyListTool::GetMetaserverInfoFromResponse() {
+    bool ret = true;
+    auto const& metaservers = response_->metaservers();
+    if (metaservers.statuscode() != curvefs::mds::topology::TOPO_OK) {
+        std::cerr << "the metaservers of cluster has error, error code is "
+                  << metaservers.statuscode() << ", code name is "
+                  << curvefs::mds::topology::TopoStatusCode_Name(
+                         metaservers.statuscode())
+                  << std::endl;
+        ret = false;
+    } else {
+        for (auto const& i : metaservers.metaserverinfos()) {
+            metaserverId2MetaserverInfo.insert(
+                std::pair<mds::topology::MetaServerIdType, MetaserverInfoType>(
+                    i.metaserverid(), MetaserverInfoType(i)));
+            if (serverId2ServerInfo.find(i.serverid()) !=
+                serverId2ServerInfo.end()) {
+                serverId2ServerInfo[i.serverid()].second.emplace_back(
+                    i.metaserverid());
+            } else {
+                errorOutput_ << "metaserver:" << i.metaserverid()
+                             << " has error: serverId:" << i.serverid()
+                             << " can't be found." << std::endl;
+                ret = false;
+            }
+        }
+    }
+    return ret;
+}
+
+void TopologyListTool::ShowPoolInfo(const PoolInfoType& pool) const {
+    auto const& poolInfo = pool.first;
+    std::cout << "poolId:" << poolInfo.poolid()
```

Comment:
Maybe use a stringbuf and cout once is better.

Reply:
> Maybe use a stringbuf and cout once is better.

fix